### PR TITLE
Pluggable profiles and OciProfile as a specific case

### PR DESCRIPTION
**Pull request by n4ss**
Implementing this proposal: https://github.com/docker/libentitlement/issues/22

`Profile` becomes an abstract interface, implemented by `OciProfile`. Type assertions and conversion have been added where needed.

### CODE REVIEW

File: defaults/network-ents.go [hunk 1]
```diff
@@ -77,17 +87,26 @@ func networkNoneEntitlementEnforce(profile *secProfile.Profile) (*secProfile.Pro
  * - Blocked syscalls:
  * 	sethostname, setdomainname, setsockopt(SO_DEBUG)
  */
-func networkUserEntitlementEnforce(profile *secProfile.Profile) (*secProfile.Profile, error) {
+func networkUserEntitlementEnforce(profile secProfile.Profile) (secProfile.Profile, error) {
+	if profile.GetType() != secProfile.OciProfileType {
```
Comment (riyazdf) [inline defaults/network-ents.go]: can be a followup: feels like all of this profile type checking and type conversion could be factored out to a helper function, ex:
```
func ociProfileConversionCheck(profile secProfile.Profile, entitlementID) (secProfile.OciProfile, error) {
```

since you have this code copied over multiple times

File: libentitlement.go [hunk 2]
```diff
@@ -11,14 +11,14 @@ import (
 )
 
 type EntitlementsManager struct {
-	profile         *secprofile.Profile
+	profile         secprofile.Profile
```
Comment (riyazdf) [inline libentitlement.go]: I think this is fine, but what's the reason for this change?

File: security-profile/oci-profile.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+/* Add a list of capabilities if not present to all capability masks */
+func (p *OciProfile) AddCaps(capsToAdd ...string) {
+	for _, cap := range capsToAdd {
+		p.Oci.Process.Capabilities.Bounding = addCapToList(p.Oci.Process.Capabilities.Bounding, cap)
+		p.Oci.Process.Capabilities.Effective = addCapToList(p.Oci.Process.Capabilities.Effective, cap)
+		p.Oci.Process.Capabilities.Inheritable = addCapToList(p.Oci.Process.Capabilities.Inheritable, cap)
+		p.Oci.Process.Capabilities.Permitted = addCapToList(p.Oci.Process.Capabilities.Permitted, cap)
+
+		// Should be updated automatically if the previous masks are set
+		p.Oci.Process.Capabilities.Ambient = addCapToList(p.Oci.Process.Capabilities.Ambient, cap)
```
Comment (riyazdf) [inline security-profile/oci-profile.go]: can be a followup: are we sure we want to add the cap to all lists?  I know runc is very careful about adding caps to various sets, especially ambient. I can help investigate, I need to brush up on this.

File: security-profile/security_profile.go [hunk 4]
```diff
@@ -1,193 +1,9 @@
...
-		syscallsWithNoArgsToBlock[syscallsToBlock] = []specs.LinuxSeccompArg{}
-	}
-
-	p.BlockSyscallsWithArgs(syscallsWithNoArgsToBlock)
+/* Profile is an abstract interface which represents security profiles.
+ * Each security profile has its own type and its own API as needs
+ * may vary across different profile formats.
+ */
+type Profile interface {
+	GetType() string
```
Comment (riyazdf) [inline security-profile/security_profile.go]: :+1: we can go even further with defining a `ProfileType`, similar how we handle `GUN` in notary

```
type ProfileType string
```

Comment (riyazdf) [inline security-profile/security_profile.go]: Also - I wonder if there are common APIs we want to pull out into `Profile` for now?  At the very least, `GetType`?

File: security-profile/oci-profile.go (hunk 3 above)
Comment (n4ss) [inline security-profile/oci-profile.go]: As far as I understood the code I dug in, `Ambient` is supposed to be updated automatically and we do add caps to every set besides Ambient in our setup of the default OCI config: https://github.com/moby/moby/blob/930e6896685f614d8f15ed10207596b1a151cbc8/oci/defaults.go#L121

File: defaults/network-ents.go [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package defaults
 
 import (
+	"fmt"
```
Comment (n4ss) [inline defaults/network-ents.go]: my bad, `gofmt` regorganized this its own way.

File: defaults/network-ents.go (hunk 1 above)
Comment (n4ss) [inline defaults/network-ents.go]: Was on the todo-list, will do 😄

File: defaults/network-ents.go (hunk 5 above)
Comment (riyazdf) [inline defaults/network-ents.go]: I think `gofmt` just only enforces alphabetical order for each block, but does not reorganize into multiple blocks

File: libentitlement.go (hunk 2 above)
Comment (n4ss) [inline libentitlement.go]: Seems like golang best practices advise to keep non-pointer interfaces as it's equivalent to a concrete struct reference.

File: security-profile/security_profile.go (hunk 4 above)
Comment (n4ss) [inline security-profile/security_profile.go]: Can you elaborate a bit on this please? Moving the `GetType()` implementation as a method returning a string initialized at the concrete type instantiation ?

Comment (n4ss) [inline security-profile/security_profile.go]: Other than getting the profile type, I'm not sure yet what we want here.

File: defaults/network-ents.go (hunk 5 above)
Comment (n4ss) [inline defaults/network-ents.go]: fixed.

File: defaults/network-ents.go (hunk 1 above)
Comment (n4ss) [inline defaults/network-ents.go]: https://github.com/docker/libentitlement/issues/27

File: security-profile/security_profile.go (hunk 4 above)
Comment (riyazdf) [inline security-profile/security_profile.go]: I think the profile type is the only common API at present to this LGTM.

I was suggesting that we could strongly type the profile type string by making it a custom type `type ProfileType string` like we do for GUNs in Notary

File: security-profile/oci-profile.go (hunk 3 above)
Comment (riyazdf) [inline security-profile/oci-profile.go]: ok nice find!  Works for me :)

**Comment by n4ss** [comment]
woops, I merged https://github.com/docker/libentitlement/pull/29 but it generated conflicts.

**Comment by n4ss** [comment]
This PR is updated with conflict solving in #30 and got merged.
